### PR TITLE
refactor(agent): Extract methods from Agent class

### PR DESCRIPTION
**Pull request by yangkyeongmo**
### Background

- This PR extracts possible methods from `Agent` class.
- While working on https://github.com/Significant-Gravitas/Auto-GPT/pull/3947, I frequently met merge conflicts. The feature was simple wrapping using async, but changes in position caused the merge conflict.
- This PR aims to decrease the complexity of agent code.
- Since it seems major architecture refactoring is in progress, this PR aims to move existing code without wrapping them in other objects, 

### Changes

- No feature changes are expected.
- There is a moved line, I'll make a comment about it.

### Documentation

- In docstrings

### Test Plan

Tested on random prompt.

```
WARNING:  You are running on `refactor/agent` branch - this is not a supported branch.
Welcome back!  Would you like me to return to being ResearchGPT?
  Asking user via keyboard...
Continue with the last settings?
Name:  ResearchGPT
Role:  an autonomous agent designed to assist in gathering and analyzing information for research purposes. It utilizes advanced algorithms and natural language processing to provide accurate and comprehensive reports.
Goals: ['Conduct thorough research on AutoGPT, including its history, features, and capabilities, to provide a detailed report.', 'Utilize various sources, including academic journals, industry reports, and online forums, to gather relevant information.', "Analyze the gathered data to identify patterns, trends, and insights that can be used to improve the report's accuracy and usefulness.", 'Ensure that the report is well-organized, easy to understand, and visually appealing, with clear headings, subheadings, and graphics.', 'Provide the report in a timely manner, with regular updates on progress and any potential issues that may arise.']
API Budget: infinite
Continue (y/n): y
ResearchGPT  has been created with the following details:
Name:  ResearchGPT
Role:  an autonomous agent designed to assist in gathering and analyzing information for research purposes. It utilizes advanced algorithms and natural language processing to provide accurate and comprehensive reports.
Goals: 
-  Conduct thorough research on AutoGPT, including its history, features, and capabilities, to provide a detailed report.
-  Utilize various sources, including academic journals, industry reports, and online forums, to gather relevant information.
-  Analyze the gathered data to identify patterns, trends, and insights that can be used to improve the report's accuracy and usefulness.
-  Ensure that the report is well-organized, easy to understand, and visually appealing, with clear headings, subheadings, and graphics.
-  Provide the report in a timely manner, with regular updates on progress and any potential issues that may arise.
Using memory of type:  LocalCache
Using Browser:  chrome
 THOUGHTS:  I think we should start by researching AutoGPT to provide a detailed report on its history, features, and capabilities. We can use various sources such as academic journals, industry reports, and online forums to gather relevant information. I will need to save important information to files since my short-term memory is limited. I should also aim to complete tasks in the least number of steps to be efficient.
REASONING:  Researching AutoGPT will provide us with a solid foundation to build our report on. Using various sources will ensure that we have a comprehensive understanding of the topic. Saving important information to files will help me avoid losing any crucial data. Completing tasks in the least number of steps will help us be efficient and save time.
PLAN: 
-  Research AutoGPT
-  Use various sources to gather information
-  Save important information to files
-  Complete tasks in the least number of steps
CRITICISM:  I need to ensure that I am not missing any important information while researching and that the sources I use are reliable.
NEXT ACTION:  COMMAND = browse_website ARGUMENTS = {'url': 'https://autogpt.com/', 'question': 'What is AutoGPT and what are its features?'}
  Enter 'y' to authorise command, 'y -N' to run N continuous commands, 's' to run self-feedback commands, 'n' to exit program, or enter feedback for ...
  Asking user via keyboard...
Input:y
-=-=-=-=-=-=-= COMMAND AUTHORISED BY USER -=-=-=-=-=-=-= 
[WDM] - Downloading: 100%|███████████████████████████████████████████████████████████████████| 6.98M/6.98M [00:00<00:00, 9.78MB/s]
  Text length: 4069 characters
  Adding chunk 1 / 1 to memory
  Summarizing chunk 1 / 1 of length 4069 characters, or 1079 tokens
DOUBLE CHECK CONFIGURATION  Please double check that you have setup a PAID OpenAI API Account. You can read more here: https://docs.agpt.co/setup/#getting-an-api-key
  Added chunk 1 summary to memory, of length 304 characters
  Summarized 1 chunks.
SYSTEM:  Command browse_website returned: Answer gathered from website: The text mentions AutoGPT.com and CarGPT as a project that is open to strong investors and technology transfer agreements. No specific features of AutoGPT or CarGPT are mentioned. The text also warns against fake websites and provides a form to contact the private owner to make an offer. Links: ['\n \n\n Listed By\n \n\n Private Owner\n \n\n (https://autogpt.com/)', '\nDomain Parking by Domain.io\n (https://domain.io)']
 THOUGHTS:  Since the website did not provide much information on AutoGPT's features, we should look for other sources. We can search for academic journals and industry reports to gather more information. We should also consider using GPT-3.5 powered agents to delegate simple tasks such as summarizing articles or generating images. I will save all the gathered information to files to avoid losing any important data.
REASONING:  Searching for academic journals and industry reports will provide us with more detailed information on AutoGPT's features. Using GPT-3.5 powered agents will help us delegate simple tasks and save time. Saving all the gathered information to files will help us avoid losing any important data.
PLAN: 
-  Search for academic journals and industry reports
-  Consider using GPT-3.5 powered agents to delegate simple tasks
-  Save all the gathered information to files
CRITICISM:  We need to ensure that the sources we use are reliable and that the information we gather is relevant to our report.
NEXT ACTION:  COMMAND = execute_shell ARGUMENTS = {'command_line': 'pip install openai'}
  Enter 'y' to authorise command, 'y -N' to run N continuous commands, 's' to run self-feedback commands, 'n' to exit program, or enter feedback for ...
  Asking user via keyboard...
Input:n
  Exiting...
Press any key to continue...
```

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: autogpt/agent/agent.py [hunk 1]
```diff
@@ -294,6 +169,142 @@ def start_interaction_loop(self):
...
                 )
 
+    def _get_json_assistant_reply(self, assistant_reply, cfg):
+        assistant_reply_json = fix_json_using_multiple_techniques(assistant_reply)
+        for plugin in cfg.plugins:
+            if not plugin.can_handle_post_planning():
+                continue
+            assistant_reply_json = plugin.post_planning(assistant_reply_json)
+        if assistant_reply_json != {}:
+            validate_json(assistant_reply_json, LLM_DEFAULT_RESPONSE_FORMAT)
```
Comment (yangkyeongmo) [inline autogpt/agent/agent.py:179]: These lines were originally near getting command and arguments.
These looked more fit with getting assistant reply.

**Comment by Boostrix** [comment]
I believe I saw at least one PR where someone started cleaning up the main loop of the agent to move stuff into methods, you may want to search open PRs to see for yourself (sorry if that's also you, I don't remember TBH)

File: autogpt/agent/agent.py [hunk 2]
```diff
@@ -119,111 +119,25 @@ def start_interaction_loop(self):
...
-            logger.typewriter_log(
-                "NEXT ACTION: ",
-                Fore.CYAN,
-                f"COMMAND = {Fore.CYAN}{command_name}{Style.RESET_ALL}  "
-                f"ARGUMENTS = {Fore.CYAN}{arguments}{Style.RESET_ALL}",
-            )
+                else:
+                    self._print_assistant_thoughts(
+                        arguments, assistant_reply_json, cfg, command_name
+                    )
```
Comment (yangkyeongmo) [inline autogpt/agent/agent.py:133]: This part used to be in same try-except block with get_command.
Getting command and printing thoughts seems to be separate behavior so it looks they could be split.
Also this looked like it should only be executed if there isn't any error, so I moved it to else block.

**Review by yangkyeongmo — COMMENTED** [review]
Here's complementary comment for reviewers. Please have a look and any feedbacks are welcome!

**Comment by yangkyeongmo** [comment]
@Boostrix Cool! I'm sure that's not me, thanks for telling me this.

**Comment by Boostrix** [comment]
> @Boostrix Cool! I'm sure that's not me, thanks for letting me know.

I believe, it's this one: https://github.com/Significant-Gravitas/Auto-GPT/pull/1264

**Comment by yangkyeongmo** [comment]
@Boostrix Thanks for the link, I appreciate your help.

After reading through the discord and re-arch, it seems this kind of refactoring and features could be easily replaced after next architecture change.

**Comment by Boostrix** [comment]
yes, agreed: to be honest, there seems to be a little disjoint when it comes to what's being communicated behind the scenes and what's happening here in the repo versus what's publicly annnounced - the wiki actually says that discord should be for short-term stuff, versus large-scale design/arch efforts should be better formalized using the wiki/gitub:

 https://github.com/Significant-Gravitas/Auto-GPT/wiki/Discord
> GitHub is great for more-persistent information.

That's obviously not what's currently happening unfortunately.
But not knowing what people are working on is obviously a huge problem - some features/ideas here have up to 5 different people/PRs ...

Also, a re-arch effort without mentioning tasks related to the agent/tasking itself and to the challenge of dealing with persistency/checkpointing, isn't taking into account the majority of questions and RFEs on here.

Maybe things will improve once the dust settles.

**Comment by yangkyeongmo** [comment]
@Boostrix Yeah, I guess I shouldn't invest more time in this. Closing PR.

**Comment by Boostrix** [comment]
cleaning up the loop is a good idea though - maybe it's better to talk to the folks in charge, which currently seems to require discord though

**Pull request closed without merging**